### ISSUE TITLE
Unnecessary try/catch in __awaiter

### ISSUE THREAD

**Issue by wnayes** [Microsoft/TypeScript#24819]
The `__awaiter` helper includes try/catch blocks within the `fulfilled` and `rejected` functions. When an error is caught by these, it calls `reject` with the error. 

**TypeScript Version:**  3.0.0-dev.201xxxxx

```ts
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
```

This seems unnecessary, as per standard, the Promise implementation should do the rejection if an error is thrown.

From Promises/A+:

> `promise2 = promise1.then(onFulfilled, onRejected);`
> 2.2.7.2  If either onFulfilled or onRejected throws an exception e, promise2 must be rejected with e as the reason.

And the reason I write this issue is that, not only does it appear unnecessary, but it prevents my custom Promise implementation from recognizing the difference between a true runtime error and an Error object that was created and passed to reject. This affects my ability to log errors as I'd like.

**Expected behavior:**

Allow the Promise implementation to be exposed to any errors and reject the promise.

```ts
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { step(generator.next(value)); }
        function rejected(value) { step(generator["throw"](value)); }
        function step(result) { resolve(result.done ? result.value : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected)); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
```

_Edit_: As I look at this more, I realize it maybe requires a little more than just removing the try/catch statements, as the promises weren't directly chained together. I made a subtle change to `step` which I think maybe chains them all together.

**Comment by rbuckton** [comment]
The helper is written to align with the ECMAScript specification for async functions and the evaluation behavior of [`await`](https://tc39.github.io/ecma262/#await), not the Promise/A+ specification. Although Promise/A+ specifies the chaining behavior, the ECMAScript specification _does not require_ the chaining behavior when adopting the state of a foreign promise-like (such as jQuery's deferred). The only requirement for an operand to `await` that is "promise-like", is that it has a callable `then` method that accepts two callbacks. The ECMAScript specification does not observe or use the return value of `then`.
